### PR TITLE
Autosetting clearing state#127

### LINKED ISSUE #127: ReleaseClearingStateSummary in project/detail
1.  Release Clearing State Summary in project-details view misses the 5th field: number of approved releases.
2. If setting the clearing state of a release to "new" manually, then the release clearing state summary of the project is updated (in project view and project detail), but in the wrong way: instead of increasing the number of new releases, the number of releases "under clearing by project team" is increased.

Reason: the release holds also a map " ClearingTeamToFossologyStatus", where for each team the fossology status is listed. This map isn't updated when the clearing state is set manually.

Solution: the clearing state of a release should not be set manually, but should be automatically set.  (Already done for "report available" and "approved". Todo: for "under clearing"). 
The map  " ClearingTeamToFossologyStatus" can probably vanish if clearingRequests are introduced, see #128.

### PR DESCRIPTION
**Pull request by alex-evo**
completed automatic management of release clearing state and removed the option of setting it manually.
closes #127

### CODE REVIEW

**Comment by adrian-evo** [comment]
Tested and it is working as expected.

File: backend/src/src-fossology/src/main/java/com/siemens/sw360/fossology/handler/FossologyFileHandler.java [hunk 1]
```diff
@@ -28,10 +28,7 @@
 import org.springframework.stereotype.Component;
 
 import java.io.InputStream;
-import java.util.Collection;
-import java.util.Map;
-import java.util.Objects;
-import java.util.Set;
+import java.util.*;
```
Comment (mcjaeger) [inline backend/src/src-fossology/src/main/java/com/siemens/sw360/fossology/handler/FossologyFileHandler.java]: (do not change that, just a remark) actually I like identifying individual objects more since the loaded objects are less? - but after years I am not sure about it any more and one must consistently doing this to have this as an effect.

Comment (alex-evo) [inline backend/src/src-fossology/src/main/java/com/siemens/sw360/fossology/handler/FossologyFileHandler.java]: That's a readability optimization that IntelliJ did for me.
`import` is a compile-time instruction. It makes no difference on the compiled byte-code. In particular, it does not affect how many classes are loaded at runtime. You can have your code without imports at all, if you like to spell out fully qualified class names every time, or you can import all the packages with an asterisk import.

File: backend/src/src-fossology/src/main/java/com/siemens/sw360/fossology/handler/FossologyFileHandler.java [hunk 2]
```diff
@@ -180,6 +179,8 @@ public Release getStatusInFossology(String releaseId, User user, String clearing
                     updateInDb = false;
                 }
             }
+            Optional<FossologyStatus> maxStatus = nullToEmptyMap(release.getClearingTeamToFossologyStatus()).values().stream().max(FossologyStatus::compareTo);
```
Comment (maximilianhuber) [inline backend/src/src-fossology/src/main/java/com/siemens/sw360/fossology/handler/FossologyFileHandler.java]: newlines would make this line of code more readeable, i.e.
```
            Optional<FossologyStatus> maxStatus = nullToEmptyMap(release.getClearingTeamToFossologyStatus())
                .values()
                .stream()
                .max(FossologyStatus::compareTo);
```

It might also be good to encapsulate this in a function.

File: backend/src/src-fossology/src/main/java/com/siemens/sw360/fossology/handler/FossologyFileHandler.java [hunk 3]
```diff
@@ -191,6 +192,28 @@ public Release getStatusInFossology(String releaseId, User user, String clearing
         return release;
     }
 
+    private void updateReleaseClearingState(Release release, FossologyStatus fossologyStatus) {
+        updateReleaseClearingState(release, Optional.of(fossologyStatus));
+    }
+    private void updateReleaseClearingState(Release release, Optional<FossologyStatus> fossologyStatus) {
+        Optional<ClearingState> newClearingState = mapFossologyStatusToClearingState(fossologyStatus);
```
Comment (maximilianhuber) [inline backend/src/src-fossology/src/main/java/com/siemens/sw360/fossology/handler/FossologyFileHandler.java]: if you replace this with a call of the form
```
    Optional<ClearingState> newClearingState = fossologyStatus.flatMap(...)
```
the function `mapFossologyStatusToClearingState` would become
```
    private Optional<ClearingState> mapFossologyStatusToClearingState(FossologyStatus fossologyStatus) {
```
and you do not have to extract the content of the optional manually.

File: frontend/sw360-portlet/src/main/java/com/siemens/sw360/portal/portlets/components/ComponentPortletUtils.java [hunk 4]
```diff
@@ -69,6 +65,9 @@ public static void updateReleaseFromRequest(PortletRequest request, Release rele
                         release.setReleaseIdToRelationship(new HashMap<String, ReleaseRelationship>());
                     updateLinkedReleaesFromRequest(request, release.releaseIdToRelationship);
                     break;
+                case CLEARING_STATE:
+                    // skip setting CLEARING_STATE. it is supposed to be set only programmatically, never from user input.
```
Comment (maximilianhuber) [inline frontend/sw360-portlet/src/main/java/com/siemens/sw360/portal/portlets/components/ComponentPortletUtils.java]: A short question:

 Doesn't this make it harder, for users which want to use SW360 without FOSSology?

@mcjaeger: what do you think?

File: libraries/lib-datahandler/src/main/java/com/siemens/sw360/datahandler/businessrules/ReleaseClearingStateSummaryComputer.java [hunk 5]
```diff
@@ -50,7 +50,10 @@ public static ReleaseClearingStateSummary computeReleaseClearingStateSummary(Lis
     }
 
     static Void addReleaseWithStates(ReleaseClearingStateSummary summary, ViewedState globalState, ViewedState myTeamState, ViewedState otherTeamState) {
-        if (globalState.present()) { // if the release has its clearing state field set use it and ignore the rest
+        // if the release has its clearing state field set to REPORT_AVAILABLE or APPROVED use it and ignore the rest
+        // otherwise, look at what fossology has to say
+        // if all else fails, default to NEW
+        if (globalState.present() && globalState.compareTo(ViewedState.REPORT_AVAILABLE) >= 0) {
```
Comment (maximilianhuber) [inline libraries/lib-datahandler/src/main/java/com/siemens/sw360/datahandler/businessrules/ReleaseClearingStateSummaryComputer.java]: This is exactly what `Optional.filter` does:

>**Parameters:**
predicate - a predicate to apply to the value, if present
**Returns:**
an Optional describing the value of this Optional if a value is present and the value matches the given predicate, otherwise an empty Optional

this could make the code shorter.

File: frontend/sw360-portlet/src/main/java/com/siemens/sw360/portal/portlets/components/ComponentPortletUtils.java (hunk 4 above)
Comment (maximilianhuber) [inline frontend/sw360-portlet/src/main/java/com/siemens/sw360/portal/portlets/components/ComponentPortletUtils.java]: Or is this just FOSSology related?

**Comment by maximilianhuber** [comment]
Functionality is tested and works as expected.

File: backend/src/src-fossology/src/main/java/com/siemens/sw360/fossology/handler/FossologyFileHandler.java (hunk 2 above)
Comment (alex-evo) [inline backend/src/src-fossology/src/main/java/com/siemens/sw360/fossology/handler/FossologyFileHandler.java]: Oh, yes! I hate that IDEA always reformats it to one line

File: backend/src/src-fossology/src/main/java/com/siemens/sw360/fossology/handler/FossologyFileHandler.java (hunk 3 above)
Comment (alex-evo) [inline backend/src/src-fossology/src/main/java/com/siemens/sw360/fossology/handler/FossologyFileHandler.java]: thanks for the tip

File: frontend/sw360-portlet/src/main/java/com/siemens/sw360/portal/portlets/components/ComponentPortletUtils.java (hunk 4 above)
Comment (mcjaeger) [inline frontend/sw360-portlet/src/main/java/com/siemens/sw360/portal/portlets/components/ComponentPortletUtils.java]: understand the point, but "sw360 is designed to work with fossology". Also, it could be OK, my points:
* clearing state is not really fossology dependent, but tool dependent.
* if there is a new tool, integration should be done accordingly (see also the ticket #128 )
* if there is no tool, no status shall be displayed there
* for the manual state, there is mainline state
does this sound acceptable?

File: libraries/lib-datahandler/src/main/java/com/siemens/sw360/datahandler/businessrules/ReleaseClearingStateSummaryComputer.java (hunk 5 above)
Comment (alex-evo) [inline libraries/lib-datahandler/src/main/java/com/siemens/sw360/datahandler/businessrules/ReleaseClearingStateSummaryComputer.java]: Too bad `globalState` is not `Optional`
